### ISSUE TITLE
AMP's inheritance of responder methods is wrong

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4343]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)| @zseil reported|
|-|-|
|Trac ID|trac#4343|
|Type|defect|
|Created|2010-03-03 17:33:56Z|
|Branch|https://github.com/twisted/twisted/tree/amp-locator-order-4343|

`amp.CommandLocator`'s metaclass recursively walks through the base classes to add inherited responder functions to the `_commandDispatch` dictionary, but it does that in the wrong order. The attached patch should fix this.

Attachments:

* [amp-inheritance.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22e/22e67bb39be9c466dbeef54dac9455eb149dba87/0121fe06081d81fde98fea62bf3229fb0d777a59.diff) (4336 bytes) - added by zseil on 2010-03-03 17:34:38Z - fix with tests

<details><summary>Searchable metadata</summary>

```
trac-id__4343 4343
type__defect defect
reporter__zseil zseil
priority__normal normal
milestone__ 
branch__branches_amp_locator_order_4343 branches/amp-locator-order-4343
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1267637636000000 1267637636000000
changetime__1268588511000000 1268588511000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)|@zseil removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [28731]) Branching to 'amp-locator-order-4343'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [28732]) Apply amp-inheritance.diff

Refs [#4343](https://github.com/twisted/twisted/issues/4343)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  The patch looks great.  There are a couple minor issues, like a docstring typo, which I'm going to correct, then I'll apply it to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [28733]) Update copyright dates; fix typo in new helper method docstring; also move helper method above the first test method which uses it

refs [#4343](https://github.com/twisted/twisted/issues/4343)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28734]) Make a note about how inheritance interacts with CommandLocator

refs [#4343](https://github.com/twisted/twisted/issues/4343)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28735]) News file for the fix

refs [#4343](https://github.com/twisted/twisted/issues/4343)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28736]) Replace the "should" style prose

refs [#4343](https://github.com/twisted/twisted/issues/4343)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [28737]) Merge amp-locator-order-4343

Author: zseil
Reviewer: exarkun
Fixes: [#4343](https://github.com/twisted/twisted/issues/4343)

Change `twisted.protocols.amp.CommandLocator` to collect responders in a way
which gives precedence to responders defined further from the base of the
inheritance hierarchy.  This makes them consistent with other kinds of things
you might find in a class hierarchy.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
